rtp/src/main/kotlin/org/jitsi/rtp/rtcp/rtcpfb/transport_layer_fb/ccfb/RtcpFbCcfbPacket.kt

```diff
@@ -0,0 +1,330 @@
+    // duplicate sequence numbers.
+    val packets: List<PacketInfo> = listOf(),
+    val reportTimestampCompactNtp: Long
+) {
+    private fun blockLength(): Int {
+        var totalSize = RtcpFbPacket.HEADER_SIZE + kTimestampLength
+
+        packets.groupBy { it.ssrc }.forEach { (ssrc, packets) ->
+            val blockSize = packets.size * 2
+            totalSize += kHeaderPerMediaSsrcLength + blockSize + RtpUtils.getNumPaddingBytes(blockSize)
+        }
+
+        return totalSize
+    }
+
+    fun writeTo(buf: ByteArray, offset: Int) {
+        val blockLength = blockLength()
+        val paddingBytes = RtpUtils.getNumPaddingBytes(blockLength)
+        val packetSize = blockLength + paddingBytes
+        require(buf.size - offset >= packetSize) {
+            "Buffer of size ${buf.size} with offset $offset too small for CCFB packet of size $packetSize"
+        }
+        rtcpHeader.apply {
+            packetType = TransportLayerRtcpFbPacket.PT
+            reportCount = RtcpFbCcfbPacket.FMT
+            length = RtpUtils.calculateRtcpLengthFieldValue(blockLength)
+        }.writeTo(buf, offset)
+
+        RtcpFbPacket.setMediaSourceSsrc(buf, offset, mediaSourceSsrc)
+        var position = offset + RtcpFbPacket.HEADER_SIZE
+
+        packets.groupBy { it.ssrc }.forEach { (ssrc, packets) ->
+            check(packets.size <= 16384) {
+                "Unexpected number of reports: ${packets.size}"
+            }
+
+            buf.putInt(position, ssrc.toInt())
+            position += 4
+
+            val beginSeq = packets.first().sequenceNumber
+            val numReports = packets.size
+
+            buf.putShort(position, beginSeq.toShort())
+            position += 2
+            buf.putShort(position, numReports.toShort())
+            position += 2
+
+            packets.forEachIndexed { i, packet ->
+                val expectedSeq = RtpUtils.applySequenceNumberDelta(beginSeq, i)
+                check(packet.sequenceNumber == expectedSeq) {
+                    "Sequence number for report $i of $ssrc is wrong " +
+                        "(expected $expectedSeq == $beginSeq + $i, got ${packet.sequenceNumber})"
+                }
+                val packetInfo = if (packet is ReceivedPacketInfo) {
+                    0x8000 or to2BitEcn(packet.ecn) or to13BitAto(packet.arrivalTimeOffset)
+                } else {
+                    0
+                }
+                buf.putShort(position, packetInfo.toShort())
+                position += 2
+            }
+            repeat(RtpUtils.getNumPaddingBytes(2 * numReports)) {
+                buf[position++] = 0x00
+            }
+        }
+
+        buf.putInt(position, reportTimestampCompactNtp.toInt())
+        position += 4
+        assert(position - offset == packetSize)
+    }
+
+    fun build(): RtcpFbCcfbPacket {
+        val blockLength = blockLength()
+        val packetSize = blockLength + RtpUtils.getNumPaddingBytes(blockLength)
```

Comment:
I think the padding is guaranteed 0. In the parsing logic you assume so, reading the timestamp from the very end.